### ISSUE TITLE
Feature request: Unsafe API

### ISSUE THREAD

**Issue by svenknobloch** [smol-rs/async-lock#17]
Would it be possible to add an unsafe API to the lock types (`Mutex/RwLock`) so that users can lock/unlock without returning a guard?

The `parking_lot` crate has something similar to this with unsafe data access, `mem::forget` and [`force_unlock`](https://docs.rs/lock_api/0.4.2/lock_api/struct.Mutex.html#method.force_unlock)

I'd imagine usage along the lines of:
```rust
let lock;
mem::forget(lock.lock().await);

unsafe {
    let value = lock.raw();
    // Use value
    lock.unsafe_unlock(); 
}
```

Alternatively unsafe methods to lock without returning a guard and retrieve a guard without locking would also work (might even be nicer API-wise).
```rust
let lock;
unsafe {
    lock.lock_raw();
    let guard = lock.raw_guard();
    drop(guard);
}
```

My use case for this is to automatically lock a set of locks in `TypeId` order which requires a homogenous type (i.e. something like `&mut dyn Future<Output = ()>`) for sorting the array of locks.

**Comment by zeenix** [comment]
@svenknobloch Sorry for the very late response here but I had kept this tab open from the day you crated this issue but forgot. :) Before we can look into your great suggestions, I wanted to ask about the use case here? What would convince me much more easily about this is if `std::sync` locks had any plans of providing such API. In general, I think people use locks to keep things safe so if they really want to bypass the safeties these locks provide, they might as well not use a lock?

**Comment by svenknobloch** [comment]
No worries 😄 

The use case is to allow for locking in a well defined order to prevent deadlocks.
The proposed API would allow for locking and retrieving the guard independently of one another, which allows for locking in a different order than creating the guards.
The reason this would be helpful is because with tuples of locks (i.e. `(Mutex<A>, Mutex<B>)`) the order of the output, which are the guards in this case, is given (`(Mutex<A>, Mutex<B>) -> (MutexGuard<A>, MutexGuard<B>)`) but the locking order needs to be done differently. (so that `(Mutex<A>, Mutex<B>).lock()` and `(Mutex<B>, Mutex<A>).lock()` don't potentially deadlock)

Also, I did revisit this just now, and ended up finding a way to do this without needing the proposed API using some other unsafe stuff (`MaybeUninit` and manual initialization).
Since the proposed API is unsafe as well, I don't think having it would provide a much cleaner solution.

Afaik, there are no plans to add such an API to the `std` locks.
Since I found another way to do this without the proposed API and it the `std` locks don't provide anything similar, I think it's probably safe to just close this.
